### PR TITLE
Completed BattleMenu

### PR DESCRIPTION
**Pull request by m-grigoriu**
- Finished implementation of BattleMenu.
- BattleMenu is designed to be launched using Application.launch(). One BattlePresenter instance should exist before BattleMenu is ever launched.
- Modified BattlePresenter to work with BattleMenu better, and to allow unit testing.
- Modified BattlePresenter and Battle related entities to remove yellow warnings.

### CODE REVIEW

File: src/main/java/com/mg105/user_interface/BattleMenu.java [hunk 1]
```diff
@@ -101,91 +142,289 @@ public void start(Stage primaryStage) throws Exception {
...
+                moveOne.setVisible(true);
+                grid.add(moveTwo, 3, 16);
+                moveTwo.setVisible(true);
+
+                moveOne.setText(moveNames[0] + "\n Hp: " + moveStats[0] + ", Dmg: " + moveStats[1]);
+                moveTwo.setText(moveNames[1] + "\n Hp: " + moveStats[2] + ", Dmg: " + moveStats[3]);
+            } else if(opponentNames[0].equals(moving) || opponentNames[1].equals(moving) ||
+                opponentNames[2].equals(moving) || opponentNames[3].equals(moving)) {
+                //If opponent moved, re-enable nextRound button
+                nextRound.setDisable(false);
```
Comment (6167656e74323431) [inline src/main/java/com/mg105/user_interface/BattleMenu.java:259]: Is there really no better way then the `if-else-if-else-if-else` spam? Surely at this point it merits a refactor. (we can decide to do this later)

**Review by 6167656e74323431 — CHANGES_REQUESTED** [review]
Surely a refactor is called for at this point!

File: src/main/java/com/mg105/interface_adapters/BattlePresenter.java [hunk 2]
```diff
@@ -65,8 +64,9 @@ public void createEncounter() {
             partyNames[i] = party.get(i).getName();
             opponentNames[i] = opponents.get(i).getName();
         }
-
-        this.view.setNames(partyNames, opponentNames);
+        if (view != null) {
+            this.view.setNames(partyNames, opponentNames);
+        }
```
[the remark this answers is not in the document]
Comment (m-grigoriu) [reply src/main/java/com/mg105/interface_adapters/BattlePresenter.java]: It is not. I removed the unnecessary check.

File: src/main/java/com/mg105/interface_adapters/BattlePresenter.java [hunk 3]
```diff
@@ -1,72 +1,39 @@
...
+    public BattlePresenter(BattleInteractor interactor) {
+        this.interactor = interactor;
+        this.interactor.setPresenter(this);
     }
 
     /**
      * Sets the view attribute.
      *
-     * @param view the BattleMenu instance to set the view attribute to.
+     * @param view the BattleMenuInterface instance to set the view attribute to.
```
Comment (siddharthgowda) [inline src/main/java/com/mg105/interface_adapters/BattlePresenter.java:26]: Its better practice to put @Override here when implementing methods from an interface

File: src/main/java/com/mg105/use_cases/BattleInteractor.java [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+     * @param presenter the BattlePresenterInterface instance to set the presenter attribute to.
+     */
+    public void setPresenter(BattlePresenterInterface presenter) {
+        this.presenter = presenter;
+    }
+
+    /**
+     * Creates a new encounter with random opponents and sets it as the current encounter in GameState
+     */
+    public void createEncounter() {
```
Comment (siddharthgowda) [inline src/main/java/com/mg105/use_cases/BattleInteractor.java:40]: If you have time you could use a builder pattern here to create the encounter but if you don't just mention it as something that we would do if we had more time (in a comment or something)

File: src/main/java/com/mg105/use_cases/BattleInteractor.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+        for (int i = 0; i < party.size(); ++i) {
+            partyNames[i] = party.get(i).getName();
+        }
+
+        for (int i = 0; i < opponents.size(); ++i) {
+            opponentNames[i] = opponents.get(i).getName();
+        }
+
+        if (this.presenter == null) { //Used for unit tests. Never executed during regular running of the application.
```
Comment (siddharthgowda) [inline src/main/java/com/mg105/use_cases/BattleInteractor.java:71]: You actually don't need this for unit tests. You can sort of define an abstract of unimplmented interface (Look at inventory interactor test) but not really a big deal.

**Review by siddharthgowda — APPROVED** [review]
Everything looks good. I left some comments about somethings that you could change but nothing is really mandatory (approved).

**Review by sbrad044 — APPROVED** [review]

File: src/main/java/com/mg105/user_interface/BattleMenu.java (hunk 1 above)
Comment (m-grigoriu) [reply src/main/java/com/mg105/user_interface/BattleMenu.java:279]: Noted.

File: src/main/java/com/mg105/use_cases/BattleInteractor.java (hunk 5 above)
Comment (m-grigoriu) [reply src/main/java/com/mg105/use_cases/BattleInteractor.java:71]: Thank you.

File: src/main/java/com/mg105/user_interface/MapDrawer.java [hunk 6]
```diff
@@ -49,7 +49,7 @@ public MapDrawer(@NotNull RoomInterpreter interpreter) {
         tiles.put(TileType.WALL, new Image(Objects.requireNonNull(RoomUpdater.class.getResourceAsStream("/tiles/wall.png"))));
         tiles.put(TileType.EXIT, new Image(Objects.requireNonNull(RoomUpdater.class.getResourceAsStream("/tiles/exit.png"))));
         tiles.put(TileType.CHEST, new Image(Objects.requireNonNull(RoomUpdater.class.getResourceAsStream("/tiles/chest.png"))));
-        tiles.put(TileType.PLAYER, new Image(Objects.requireNonNull(RoomUpdater.class.getResourceAsStream("/tiles/player.png"))));
+        tiles.put(TileType.PLAYER, new Image(Objects.requireNonNull(RoomUpdater.class.getResourceAsStream("/sprites/A.png"))));
```
Comment (m-grigoriu) [inline src/main/java/com/mg105/user_interface/MapDrawer.java:52]: Changed default character sprite. The old png file might be useless now.

File: src/main/java/com/mg105/use_cases/BattleInteractor.java [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+     * UNIMPLEMENTED
+     */
+    public void endBattle() {
+        //unimplemented, includes _removeActiveBattle and _addReward()
+    }
+
+    /**
+     * UNIMPLEMENTED
+     */
+    private void _addReward() {
```
Comment (6167656e74323431) [inline src/main/java/com/mg105/use_cases/BattleInteractor.java:281]: ...we don't do leading underscores in Java... but also it's not used so eh.

File: src/main/java/com/mg105/user_interface/MapDrawer.java (hunk 6 above)
Comment (6167656e74323431) [reply src/main/java/com/mg105/user_interface/MapDrawer.java:52]: So delete it?

**Review by 6167656e74323431 — APPROVED** [review]

**Review by siddharthgowda — APPROVED** [review]
Everything looks good for the most part. Also, the code you wrote for gamestate and application will most likely be removed when the "database" code is merged. Also, there are some instances where you make assumptions and use if statements from the fact that there are only 2 moves per character that probably violated the open/closed principle (but its probably too late to change that stuff now).

File: src/main/java/com/mg105/interface_adapters/BattlePresenter.java [hunk 8]
```diff
@@ -255,54 +116,33 @@ public ArrayList<String> retrieveTargets(int moveNum, String casterName) {
...
-            if (!target.isOpponent()) {
-                this.state.getParty().remove(target);
-                this.state.getFainted().add(target);
-            }
-        }
-
-        target.modifyDamage(m.getDamageChange());
-
-        //Currently unimplemented, will notify the view to  update the impacted character
-        this.view.updateCharacter(target.getName());
```
Comment (zhaoan12) [inline src/main/java/com/mg105/interface_adapters/BattlePresenter.java:297]: nice job cleaning up the old code

File: src/main/java/com/mg105/use_cases/RoomGetter.java [hunk 9]
```diff
@@ -50,4 +50,12 @@ public RoomGetter(@NotNull GameState state) {
 
         return new RoomLayout(chests, opponents, doorways, state.getWalkingCharacter().getCharPosition());
     }
+
+    /**
+     * Retrieves the WalkingCharacter sprite name String, returns a file name/location corresponding to the name.
+     * @return a String representing the file name/location for the selected character sprite.
+     */
+    public String getWalkingSprite() {
+        return "/sprites/" + state.getWalkingCharacter().getSpriteName() + ".png";
```
Comment (zhaoan12) [inline src/main/java/com/mg105/use_cases/RoomGetter.java:59]: interesting way to get the image

**Review by zhaoan12 — APPROVED** [review]
Cool classes, and the tests cover alot of cases.

File: src/main/java/com/mg105/user_interface/MapDrawer.java (hunk 6 above)
Comment (m-grigoriu) [reply src/main/java/com/mg105/user_interface/MapDrawer.java:52]: ok

**Pull request merged**